clustering_code/edchan-kmeans.py

```diff
@@ -0,0 +1,913 @@
+    # Stack the RGB bands to create a multi-dimensional xarray DataArray
+    rgb_values = ['B04_Red', 'B03_Green', 'B02_Blue']
+    rgb_channel = xr.DataArray(rgb_values, dims='rgb_channel', name='rgb_channel')
+    rgb_stacked = xr.concat([
+        processed_granule['red'], 
+        processed_granule['green'], 
+        processed_granule['blue']
+    ], dim=rgb_channel)
+    # Ensure that the data type is appropriate for visualization
+    # Normalize if the data is in float format, or ensure it's within [0, 255] if it's integer
+    if np.issubdtype(rgb_stacked.dtype, np.floating):
+        rgb_stacked = (rgb_stacked - rgb_stacked.min()) / (rgb_stacked.max() - rgb_stacked.min()) * 255
+    rgb_stacked = rgb_stacked.astype(np.uint8)
+    title = f'Manuel Canal-Spanish Lake: {processed_granule["date"]}'
+    rgb_plot = rgb_stacked.hvplot.rgb(
+                        x='x', y='y',
+                        xlabel='Latitude', 
+                        ylabel='Longitude',
+                        bands='rgb_channel', 
+                        data_aspect=1,
+                        title=title,
+                        crs=ccrs.UTM(16))
+    hvplot.save(rgb_plot, f'{output_folder}/{processed_granule["date"]}.png', fmt='png')
+
+# %% [markdown]
+# ### Merge and Composite Data
+# 
+# You will notice for this watershed that:
+# 1. The raster data for each date are spread across 4 granules
+# 2. Any given image is incomplete because of clouds
+# 
+# **YOUR TASK:**
+# 1. For each band:
+#     1. For each date:
+#         1. Merge all 4 granules
+#         2. Mask any negative values created by interpolating from the nodata value of -9999 (`rioxarray` should account for this, but doesn't appear to when merging. If you leave these values in they will create problems down the line)
+#     2. Concatenate the merged DataArrays along a new date dimension
+#     3. Take the mean in the date dimension to create a composite image that fills cloud gaps
+#     4. Add the band as a dimension, and give the DataArray a name
+# 2. Concatenate along the band dimension
+
+# %%
+red_das, green_das, blue_das = [], [], []
+nir_das, swir1_das, swir2_das = [], [], []
+ignore_dates = ["2023-05-22",
+                "2023-06-01",
+                "2023-06-23",
+                "2023-07-17",
+                "2023-08-02",
+                "2023-08-18",
+                "2023-08-26",
+                "2023-09-03"]
+for processed_granule in processed_granules:
+    date = processed_granule['date']
+    if date in ignore_dates:
+        continue
+    print(f"Processing date: {date}")
+    red_da = processed_granule['red']
+    green_da = processed_granule['green']
+    blue_da = processed_granule['blue']
+    nir_da = processed_granule['nir']
+    swir1_da = processed_granule['swir1']
+    swir2_da = processed_granule['swir2']
+
+    date_dimension = xr.DataArray([date], dims='date')
+    expanded_red_da = red_da.expand_dims(date=date_dimension)
```

Comment:
This is also pretty repetitive. You could write a function or method as @advythr recommends, or use a loop without a function. This is one reason why I like structuring multi-file datasets in a DataFrame -- it makes looping easy and lets me keep track of the metadata without having to create a bunch of variables.